### PR TITLE
#511 Make slash command arguments available in SB starter

### PR DESCRIPTION
**Pull request by symphony-elias**
### Ticket
#511

### Description
Make slash command arguments available in annotated methods with @Slash in SB starter

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Javadoc added or updated
- [x] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

**Comment by symphony-elias** [comment]
In this implementation, methods annotated `@Slash` with a command pattern containing arguments must have all arguments declared as method parameters in order to be registered and triggered. Is it something we want to enforce (since everything is in the `CommandContext`)?

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 1]
```diff
@@ -32,6 +37,8 @@
  * @see <a href="https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/event/EventListenerMethodProcessor.java">EventListenerMethodProcessor.java</a>
  */
 @Slf4j
+@Generated // means excluded from test coverage: error cases hard to test,
```
Comment (symphony-thibault) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:40]: Could we move this annotation on specific methods that we can't fully cover? Assuming that most of the class can still be correctly covered

**Review by symphony-thibault — APPROVED** [review]
LGTM 👍

**Review by symphony-youri — COMMENTED** [review]
Does an existing command like "/command blah" still works?

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java [hunk 2]
```diff
@@ -43,6 +48,13 @@
       "com.symphony.bdk.core."
   };
 
+  private static final ParameterNameDiscoverer PARAMETER_NAME_DISCOVERER = new DefaultParameterNameDiscoverer();
+
+  /**
+   * Key is the annotated slash method, value is the map of (parameter name, index)
+   */
+  private static final Map<Method, Map<String, Integer>> METHOD_TO_ARGUMENT_INDEXES = new HashMap<>();
```
Comment (symphony-youri) [inline symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:56]: do we use it as a cache? I would expect to find in the code something where we try to read an entry and if we don't find it we initialize, here it looks like it is used as a regular field (or maybe I missed something)

File: docs/spring-boot/core-starter.md [hunk 3]
```diff
@@ -291,6 +291,49 @@ public class SlashHello {
 By default, the `@Slash` annotation is configured to require bot mention in order to trigger the command. You can override
 this value using `@Slash#mentionBot` annotation parameter. 
 
+You can also use slash commands with arguments. To do so, the field `value` of the `@Slash` annotation must have a valid
+format as explained in the [Activity API section](../activity-api.md#Slash-command-pattern-format). 
+If the slash command pattern is valid, you will have to specify all slash arguments as method parameter with the same name and type.
+If slash command pattern or method signature is incorrect, a `warn` message will appear in your application log and
```
Comment (symphony-youri) [inline docs/spring-boot/core-starter.md:297]: or would it be better to fail the startup? at least the user is aware he has made a mistake

Comment (symphony-youri) [reply docs/spring-boot/core-starter.md:297]: mm the warning was there before... would it be a breaking change to change that....

Comment (symphony-elias) [reply docs/spring-boot/core-starter.md:297]: Yes, I would say it is a breaking change. Let's discuss that with @symphony-thibault

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java (hunk 2 above)
Comment (symphony-elias) [reply symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:56]: Yes and it is initialized line 215. I will update the implementation to make it clearer.

**Comment by symphony-elias** [comment]
> Does an existing command like "/command blah" still works?

Yes, it was tested here: symphony-bdk-core/src/test/java/com/symphony/bdk/core/activity/parsing/SlashCommandPatternTest.java
I will more tests in SlashCommandTest to make it more obvious.

File: symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java (hunk 2 above)
Comment (symphony-elias) [reply symphony-bdk-spring/symphony-bdk-core-spring-boot-starter/src/main/java/com/symphony/bdk/spring/annotation/SlashAnnotationProcessor.java:56]: Well, it seems we don't really need a cache. The DefaultNameParameterNameDiscoverer already caches some of the results (see LocalVariableTableParameterNameDiscoverer in SB)

File: docs/spring-boot/core-starter.md (hunk 3 above)
Comment (symphony-thibault) [reply docs/spring-boot/core-starter.md:297]: @symphony-elias @symphony-youri I'd keep the warning for now. We could maybe propose that as a possible improvement and let the community decide (through a GH issue)

**Pull request merged**